File: lib/services/external_search_suggestion_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

/// 外部站点首字母联想服务。
///
/// 负责聚合腾讯、爱奇艺和芒果的公开联想接口，
/// 将首字母查询统一转换为去重后的片名列表。
class ExternalSearchSuggestionService {
  /// 创建外部站点首字母联想服务。
  ///
  /// [httpClient] 允许在测试中注入假客户端，避免依赖真实网络。
  ExternalSearchSuggestionService({
    http.Client? httpClient,
  }) : _httpClient = httpClient ?? http.Client();

  /// 请求超时时间。
  static const Duration _requestTimeout = Duration(seconds: 5);

  /// HTTP 客户端。
  final http.Client _httpClient;

  /// 聚合首字母联想结果。
  ///
  /// 只有纯英文字母或数字查询才会触发外部联想，避免中文整词搜索误走首字母接口。
  Future<List<String>> fetchSuggestions(String query) async {
    final normalizedQuery = query.trim().toUpperCase();
    if (normalizedQuery.isEmpty ||
        !RegExp(r'^[A-Z0-9]+$').hasMatch(normalizedQuery)) {
      return <String>[];
    }

    _logSuggestionSummary(
      query: normalizedQuery,
      stage: '开始请求',
      platformLogs: const <_SuggestionPlatformLog>[],
      mergedSuggestions: const <String>[],
      dedupedSuggestions: const <String>[],
    );

    final platformRequests = <_SuggestionPlatformRequest>[
      _SuggestionPlatformRequest(
        platform: '腾讯视频',
        future: _fetchTencentSuggestions(normalizedQuery),
      ),
      _SuggestionPlatformRequest(
        platform: '爱奇艺',
        future: _fetchIqiyiSuggestions(normalizedQuery.toLowerCase()),
      ),
      _SuggestionPlatformRequest(
        platform: '芒果 TV',
        future: _fetchMgtvSuggestions(normalizedQuery.toLowerCase()),
      ),
    ];

    final platformLogs = <_SuggestionPlatformLog>[];
    final mergedSuggestions = <String>[];

    for (final request in platformRequests) {
      final suggestions = await request.future;
      platformLogs.add(
        _SuggestionPlatformLog(
          platform: request.platform,
          suggestions: suggestions,
        ),
      );
      mergedSuggestions.addAll(suggestions);
    }

    final dedupedSuggestions = _dedupeSuggestions(mergedSuggestions);
    _logSuggestionSummary(
      query: normalizedQuery,
      stage: '请求完成',
      platformLogs: platformLogs,
      mergedSuggestions: mergedSuggestions,
      dedupedSuggestions: dedupedSuggestions,
    );
    return dedupedSuggestions;
  }

  /// 请求腾讯视频联想词。
  Future<List<String>> _fetchTencentSuggestions(String query) async {
    try {
      final response = await _httpClient
          .post(
            Uri.parse(
              'https://actapi.video.qq.com/'
              'trpc.videosearch.smartboxServer.SugRecallHttp/GetSugHttp'
              '?vplatform=2',
            ),
            headers: const <String, String>{
              'accept': 'application/json',
              'content-type': 'application/json',
              'origin': 'https://film.qq.com',
              'referer': 'https://film.qq.com/',
              'user-agent': _defaultUserAgent,
            },
            body: json.encode(<String, dynamic>{
              'query': query,
              'page_num': 0,
              'page_size': 10,
              'scene_id': 5,
              'sug_id': 'selene_$query',
              'auth_info': <String, String>{
                'app_id': '3172',
                'app_key': 'lGhFIPeD3HsO9xEp',
              },
            }),
          )
          .timeout(_requestTimeout);

      if (response.statusCode != 200) {
        return <String>[];
      }

      final Map<String, dynamic> data =
          json.decode(response.body) as Map<String, dynamic>;
      final Map<String, dynamic> resultList =
          ((data['data'] as Map<String, dynamic>?)?['result_list']
                  as Map<String, dynamic>?) ??
              const <String, dynamic>{};
      final List<dynamic> rawItems =
          (resultList['item_list'] as List<dynamic>?) ?? <dynamic>[];

      final suggestions = <String>[];
      for (final rawItem in rawItems) {
        if (rawItem is! Map<String, dynamic>) {
          continue;
        }
        final view = rawItem['view'];
        if (view is! Map<String, dynamic>) {
          continue;
        }
        final lines = view['lines'];
        if (lines is! List<dynamic> || lines.isEmpty) {
          continue;
        }
        final firstLine = lines.first;
        if (firstLine is! Map<String, dynamic>) {
          continue;
        }
        final text = firstLine['text']?.toString() ?? '';
        final cleanText = text.replaceAll(RegExp(r'</?em>'), '').trim();
        if (cleanText.isNotEmpty) {
          suggestions.add(cleanText);
        }
      }
      return suggestions;
    } catch (_) {
      return <String>[];
    }
  }

  /// 请求爱奇艺联想词。
  Future<List<String>> _fetchIqiyiSuggestions(String query) async {
    try {
      final response = await _httpClient
          .get(
            Uri.parse(
              'https://mesh.if.iqiyi.com/portal/lw/search/searchKeyWord'
              '?key=$query'
              '&version=17.054.25384'
              '&deviceId=6ab07a4f0966c704610f0448494b67cb'
              '&appMode='
              '&os='
              '&pcv=17.054.25384',
            ),
            headers: const <String, String>{
              'accept': '*/*',
              'origin': 'https://www.iqiyi.com',
              'referer': 'https://www.iqiyi.com/',
              'user-agent': _defaultUserAgent,
            },
          )
          .timeout(_requestTimeout);

      if (response.statusCode != 200) {
        return <String>[];
      }

      final Map<String, dynamic> data =
          json.decode(response.body) as Map<String, dynamic>;
      final List<dynamic> rawItems =
          ((data['data'] as Map<String, dynamic>?)?['keyWordData']
                  as List<dynamic>?) ??
              <dynamic>[];

      return rawItems
          .whereType<Map<String, dynamic>>()
          .map((item) => item['name']?.toString().trim() ?? '')
          .where((title) => title.isNotEmpty)
          .toList();
    } catch (_) {
      return <String>[];
    }
  }

  /// 请求芒果 TV 联想词。
  Future<List<String>> _fetchMgtvSuggestions(String query) async {
    try {
      final response = await _httpClient
          .get(
            Uri.parse(
              'https://mobileso.bz.mgtv.com/pc/suggest/v1'
              '?allowedRC=1'
              '&src=mgtv'
              '&did=5bbe5b75-33e3-471c-87df-de12d62291f2'
              '&pc=1'
              '&q=$query'
              '&_support=10000000',
            ),
            headers: const <String, String>{
              'accept': 'application/json, text/plain, */*',
              'origin': 'https://www.mgtv.com',
              'referer': 'https://www.mgtv.com/',
              'user-agent': _defaultUserAgent,
            },
          )
          .timeout(_requestTimeout);

      if (response.statusCode != 200) {
        return <String>[];
      }

      final Map<String, dynamic> data =
          json.decode(response.body) as Map<String, dynamic>;
      final List<dynamic> rawItems =
          ((data['data'] as Map<String, dynamic>?)?['suggest']
                  as List<dynamic>?) ??
              <dynamic>[];

      return rawItems
          .whereType<Map<String, dynamic>>()
          .map((item) => item['title']?.toString().trim() ?? '')
          .where((title) => title.isNotEmpty)
          .toList();
    } catch (_) {
      return <String>[];
    }
  }

  /// 对聚合后的联想结果做有序去重。
  List<String> _dedupeSuggestions(List<String> suggestions) {
    final orderedSuggestions = <String>[];
    final seenSuggestions = <String>{};
    for (final suggestion in suggestions) {
      final normalizedSuggestion = suggestion
          .replaceAll(RegExp(r'\s+'), ' ')
          .trim();
      if (normalizedSuggestion.isEmpty ||
          !seenSuggestions.add(normalizedSuggestion)) {
        continue;
      }
      orderedSuggestions.add(normalizedSuggestion);
    }
    return orderedSuggestions;
  }

  /// 打印首字母联想聚合统计。
  ///
  /// 方便在控制台快速查看各平台命中词、原始总数和最终去重结果。
  void _logSuggestionSummary({
    required String query,
    required String stage,
    required List<_SuggestionPlatformLog> platformLogs,
    required List<String> mergedSuggestions,
    required List<String> dedupedSuggestions,
  }) {
    final buffer = StringBuffer()
      ..writeln('[首字母联想] $stage')
      ..writeln('[首字母联想] 查询词: $query')
      ..writeln('[首字母联想] 接入平台数: ${platformLogs.length}');

    for (final platformLog in platformLogs) {
      buffer
        ..writeln(
          '[首字母联想] ${platformLog.platform} 命中数量: ${platformLog.suggestions.length}',
        )
        ..writeln(
          '[首字母联想] ${platformLog.platform} 命中词: ${platformLog.suggestions.join(' | ')}',
        );
    }

    if (platformLogs.isNotEmpty) {
      buffer
        ..writeln('[首字母联想] 聚合前总数量: ${mergedSuggestions.length}')
        ..writeln('[首字母联想] 去重后数量: ${dedupedSuggestions.length}')
        ..writeln(
          '[首字母联想] 最终联想词: ${dedupedSuggestions.join(' | ')}',
        );
    }

    debugPrint(buffer.toString().trimRight());
  }

  /// 默认浏览器请求头 UA。
  static const String _defaultUserAgent =
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) '
      'AppleWebKit/537.36 (KHTML, like Gecko) '
      'Chrome/148.0.0.0 Safari/537.36';
}

/// 单个平台的联想请求描述。
class _SuggestionPlatformRequest {
  /// 创建单个平台联想请求。
  const _SuggestionPlatformRequest({
    required this.platform,
    required this.future,
  });

  /// 平台名称。
  final String platform;

  /// 平台请求任务。
  final Future<List<String>> future;
}

/// 单个平台的联想结果日志。
class _SuggestionPlatformLog {
  /// 创建单个平台联想结果日志。
  const _SuggestionPlatformLog({
    required this.platform,
    required this.suggestions,
  });

  /// 平台名称。
  final String platform;

  /// 当前平台返回的联想词。
  final List<String> suggestions;
}

File: re-android/core-network/src/main/java/org/moontechlab/selene/tv/core/network/ExternalSearchSuggestionService.kt
package org.moontechlab.selene.tv.core.network

import com.google.gson.JsonParser
import java.util.concurrent.TimeUnit
import java.util.logging.Logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.Proxy

/**
 * 外部站点首字母联想服务。
 *
 * 对齐 Flutter [ExternalSearchSuggestionService]：
 * 聚合腾讯视频、爱奇艺、芒果 TV 公开联想接口，仅处理纯字母/数字首字母查询。
 *
 * @property client 专用于外站联想的 OkHttp 客户端（直连、短超时）。
 */
class ExternalSearchSuggestionService(
    private val client: OkHttpClient = createDefaultClient(),
) {
    /**
     * 聚合首字母联想结果。
     *
     * @param query 左侧键盘拼出的查询串。
     * @return 去重后的片名列表；非首字母查询返回空。
     */
    suspend fun fetchSuggestions(query: String): List<String> = withContext(Dispatchers.IO) {
        val normalizedQuery = query.trim().uppercase()
        if (!isInitialsQuery(normalizedQuery)) {
            return@withContext emptyList()
        }
        LOGGER.info("[首字母联想] 开始请求 query=$normalizedQuery")

        // 三端并行，总等待接近最慢一端（Flutter 串行；Kotlin 并行提速，解析规则一致）。
        val (tencent, iqiyi, mgtv) = coroutineScope {
            val tencentDeferred = async {
                runCatching { fetchTencentSuggestions(normalizedQuery) }.getOrDefault(emptyList())
            }
            val iqiyiDeferred = async {
                runCatching { fetchIqiyiSuggestions(normalizedQuery.lowercase()) }.getOrDefault(emptyList())
            }
            val mgtvDeferred = async {
                runCatching { fetchMgtvSuggestions(normalizedQuery.lowercase()) }.getOrDefault(emptyList())
            }
            Triple(tencentDeferred.await(), iqiyiDeferred.await(), mgtvDeferred.await())
        }

        LOGGER.info(
            "[首字母联想] 腾讯=${tencent.size} 爱奇艺=${iqiyi.size} 芒果=${mgtv.size}",
        )
        val merged = tencent + iqiyi + mgtv
        val deduped = dedupeSuggestions(merged)
        LOGGER.info("[首字母联想] 去重后=${deduped.size} 词=$deduped")
        deduped
    }

    /**
     * 请求腾讯视频联想词。
     */
    private fun fetchTencentSuggestions(query: String): List<String> {
        val bodyJson = """
            {
              "query":"$query",
              "page_num":0,
              "page_size":10,
              "scene_id":5,
              "sug_id":"selene_$query",
              "auth_info":{"app_id":"3172","app_key":"lGhFIPeD3HsO9xEp"}
            }
        """.trimIndent()
        val request = Request.Builder()
            .url(TENCENT_URL)
            .post(bodyJson.toRequestBody(JSON_MEDIA_TYPE))
            .header("accept", "application/json")
            .header("content-type", "application/json")
            .header("origin", "https://film.qq.com")
            .header("referer", "https://film.qq.com/")
            .header("user-agent", DEFAULT_USER_AGENT)
            .build()
        return execute(request, ::parseTencentBody)
    }

    /**
     * 请求爱奇艺联想词。
     */
    private fun fetchIqiyiSuggestions(query: String): List<String> {
        val url = "https://mesh.if.iqiyi.com/portal/lw/search/searchKeyWord" +
            "?key=$query" +
            "&version=17.054.25384" +
            "&deviceId=6ab07a4f0966c704610f0448494b67cb" +
            "&appMode=" +
            "&os=" +
            "&pcv=17.054.25384"
        val request = Request.Builder()
            .url(url)
            .get()
            .header("accept", "*/*")
            .header("origin", "https://www.iqiyi.com")
            .header("referer", "https://www.iqiyi.com/")
            .header("user-agent", DEFAULT_USER_AGENT)
            .build()
        return execute(request, ::parseIqiyiBody)
    }

    /**
     * 请求芒果 TV 联想词。
     */
    private fun fetchMgtvSuggestions(query: String): List<String> {
        val url = "https://mobileso.bz.mgtv.com/pc/suggest/v1" +
            "?allowedRC=1" +
            "&src=mgtv" +
            "&did=5bbe5b75-33e3-471c-87df-de12d62291f2" +
            "&pc=1" +
            "&q=$query" +
            "&_support=10000000"
        val request = Request.Builder()
            .url(url)
            .get()
            .header("accept", "application/json, text/plain, */*")
            .header("origin", "https://www.mgtv.com")
            .header("referer", "https://www.mgtv.com/")
            .header("user-agent", DEFAULT_USER_AGENT)
            .build()
        return execute(request, ::parseMgtvBody)
    }

    private fun execute(request: Request, parser: (String) -> List<String>): List<String> {
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                return emptyList()
            }
            val body = response.body?.string().orEmpty()
            if (body.isBlank()) {
                return emptyList()
            }
            return runCatching { parser(body) }.getOrDefault(emptyList())
        }
    }

    companion object {
        private val LOGGER = Logger.getLogger(ExternalSearchSuggestionService::class.java.name)
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
        private const val TENCENT_URL =
            "https://actapi.video.qq.com/" +
                "trpc.videosearch.smartboxServer.SugRecallHttp/GetSugHttp" +
                "?vplatform=2"
        private const val DEFAULT_USER_AGENT =
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) " +
                "AppleWebKit/537.36 (KHTML, like Gecko) " +
                "Chrome/148.0.0.0 Safari/537.36"
        private const val REQUEST_TIMEOUT_SECONDS = 5L

        /**
         * 是否为纯字母数字首字母查询。
         *
         * @param query 已 trim 的查询串。
         * @return true 表示可走外部联想。
         */
        fun isInitialsQuery(query: String): Boolean {
            return query.isNotEmpty() && query.matches(Regex("^[A-Z0-9]+$"))
        }

        /**
         * 创建默认直连短超时客户端。
         */
        fun createDefaultClient(): OkHttpClient {
            return OkHttpClient.Builder()
                .proxy(Proxy.NO_PROXY)
                .connectTimeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .readTimeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .callTimeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .build()
        }

        /**
         * 解析腾讯视频联想 JSON。
         *
         * @param body 响应体。
         * @return 片名列表。
         */
        fun parseTencentBody(body: String): List<String> {
            val root = JsonParser.parseString(body).asJsonObject
            val itemList = root
                .getAsJsonObject("data")
                ?.getAsJsonObject("result_list")
                ?.getAsJsonArray("item_list")
                ?: return emptyList()
            val suggestions = ArrayList<String>()
            for (element in itemList) {
                if (!element.isJsonObject) continue
                val lines = element.asJsonObject
                    .getAsJsonObject("view")
                    ?.getAsJsonArray("lines")
                    ?: continue
                if (lines.size() == 0) continue
                val first = lines[0]
                if (!first.isJsonObject) continue
                val text = first.asJsonObject.get("text")?.asString.orEmpty()
                val clean = text.replace(Regex("</?em>"), "").trim()
                if (clean.isNotEmpty()) {
                    suggestions += clean
                }
            }
            return suggestions
        }

        /**
         * 解析爱奇艺联想 JSON。
         *
         * @param body 响应体。
         * @return 片名列表。
         */
        fun parseIqiyiBody(body: String): List<String> {
            val root = JsonParser.parseString(body).asJsonObject
            val items = root
                .getAsJsonObject("data")
                ?.getAsJsonArray("keyWordData")
                ?: return emptyList()
            return items.mapNotNull { element ->
                if (!element.isJsonObject) return@mapNotNull null
                element.asJsonObject.get("name")?.asString?.trim()?.takeIf { it.isNotEmpty() }
            }
        }

        /**
         * 解析芒果 TV 联想 JSON。
         *
         * @param body 响应体。
         * @return 片名列表。
         */
        fun parseMgtvBody(body: String): List<String> {
            val root = JsonParser.parseString(body).asJsonObject
            val items = root
                .getAsJsonObject("data")
                ?.getAsJsonArray("suggest")
                ?: return emptyList()
            return items.mapNotNull { element ->
                if (!element.isJsonObject) return@mapNotNull null
                element.asJsonObject.get("title")?.asString?.trim()?.takeIf { it.isNotEmpty() }
            }
        }

        /**
         * 对聚合结果做有序去重。
         *
         * @param suggestions 原始列表。
         * @return 去重后的列表。
         */
        fun dedupeSuggestions(suggestions: List<String>): List<String> {
            val ordered = ArrayList<String>()
            val seen = LinkedHashSet<String>()
            for (suggestion in suggestions) {
                val normalized = suggestion.replace(Regex("\\s+"), " ").trim()
                if (normalized.isEmpty() || !seen.add(normalized)) {
                    continue
                }
                ordered += normalized
            }
            return ordered
        }
    }
}
